src/nodes/dir.rs

```diff
@@ -129,15 +136,60 @@ impl Node for Dir {
         Ok(state.attr)
     }
 
-    fn lookup(&self, _name: &OsStr) -> NodeResult<(Arc<Node>, fuse::FileAttr)> {
-        Err(KernelError::from_errno(libc::ENOENT))
+    fn lookup(&self, name: &OsStr, ids: &IdGenerator, cache: &Cache)
+        -> NodeResult<(Arc<Node>, fuse::FileAttr)> {
+        let mut state = self.state.lock().unwrap();
+
+        if let Some(node) = state.children.get(name) {
+            let refreshed_attr = node.getattr()?;
+            return Ok((node.clone(), refreshed_attr))
+        }
+
+        let (child, attr) = {
+            let path = match state.underlying_path.as_ref() {
+                Some(underlying_path) => underlying_path.join(name),
+                None => return Err(KernelError::from_errno(libc::ENOENT)),
+            };
+            let fs_attr = fs::symlink_metadata(&path)?;
+            let node = cache.get_or_create(ids, &path, &fs_attr, self.writable);
+            let attr = conv::attr_fs_to_fuse(path.as_path(), node.inode(), &fs_attr);
+            (node, attr)
+        };
+        state.children.insert(name.to_os_string(), child.clone());
+        Ok((child, attr))
     }
 
-    fn readdir(&self, reply: &mut fuse::ReplyDirectory) -> NodeResult<()> {
-        let state = self.state.lock().unwrap();
+    fn readdir(&self, ids: &IdGenerator, cache: &Cache, reply: &mut fuse::ReplyDirectory)
```

Comment:
I'm not a fan of passing down the fuse-specific thing here. It's actually confusing -- this function only calls "add" but not "ok" or "error". IMO this function should return a Vec of directory entries, or, even better, an Iterator of directory entries (which lets us implement pagination later).

Reply:
Agreed in principle. However, the reason I went with the current approach is to avoid creating an ephemeral vector for the return value, only to later iterate over it and copy all data into the kernel reply. The other problem is that this vector ought to contain only the data that `reply.add()` later needs... which means having to define another struct just for this.

For now I've just updated the description of the method to be clear on what the expectations on the `reply` are. Thoughts?